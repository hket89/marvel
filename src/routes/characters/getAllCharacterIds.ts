import { CharacterResponse } from 'src/types/marvel';

import { cache } from '../../lib/cache';
import { getCharacters } from '../../services/marvel';
import { Middleware } from '../../types/koa';

const CACHE_KEY = 'ALL_MARVEL_CHARACTER_ID' as const;
const LIMIT = 100 as const;

const getFromAPI = async (limit: number) => {
  const firstResult = await getCharacters({ params: { limit } });

  const {
    data: { total }
  } = firstResult;

  const count = Math.floor(total / limit);

  const result = (await Promise.all(
    Array.from(Array(count), (_, x) =>
      getCharacters({ params: { offset: (x + 1) * limit, limit } })
    )
  )) as CharacterResponse[];

  return [firstResult, ...result]
    .map((x) => x.data.results.map((y) => y.id))
    .flat();
};

export const getData = async () => {
  const value = cache.get<number[]>(CACHE_KEY);

  if (value) {
    return value;
  }

  const result = await getFromAPI(LIMIT);
  cache.set(CACHE_KEY, result);

  return result;
};

export const getAllCharacterIdsHandler: Middleware = async (ctx) => {
  const results = await getData();
  ctx.body = results;
};
