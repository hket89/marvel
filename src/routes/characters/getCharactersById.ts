import { CharacterResponse } from 'src/types/marvel';

import { getCharacterById } from '../../services/marvel';
import { Middleware } from '../../types/koa';
import { characterIdPathSchema } from '../../types/schema';

const massageResponse = (input: CharacterResponse) => {
  const {
    data: { results }
  } = input;
  const [{ id, name, description }] = results;

  return {
    id,
    name,
    description
  };
};

export const getCharactersByIdHandler: Middleware = async (ctx) => {
  const { id } = characterIdPathSchema.validateSync(ctx.params);

  const result = await getCharacterById(id);

  ctx.body = massageResponse(result);
};
