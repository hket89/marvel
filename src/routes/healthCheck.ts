import { Middleware } from '../types/koa';

export const healthCheckHandler: Middleware = (ctx) => {
  ctx.body = 'ok';
};
