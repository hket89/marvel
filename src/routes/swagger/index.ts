import { koaSwagger } from 'koa2-swagger-ui';

import openApi from './openapi.json';

const spec = openApi;

export const swagger = koaSwagger({
  routePrefix: '/api-docs', // host at /swagger instead of default /docs
  swaggerOptions: { spec }
});
