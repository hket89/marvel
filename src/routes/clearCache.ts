import { cache } from '../lib/cache';
import { Middleware } from '../types/koa';
import { ALL_MARVEL_CHARACTER_ID } from './characters/getAllCharacterIds';

export const clearCacheHandler: Middleware = (ctx) => {
  cache.del(ALL_MARVEL_CHARACTER_ID);
  ctx.body = 'ok';
};
