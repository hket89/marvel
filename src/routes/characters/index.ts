import Router from '@koa/router';

import { getAllCharacterIdsHandler } from './getAllCharacterIds';
import { getCharactersByIdHandler } from './getCharactersById';

export const characterRouter = new Router()
  .get('/', getAllCharacterIdsHandler)
  .get('/:id', getCharactersByIdHandler);
