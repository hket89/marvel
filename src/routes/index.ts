import Router from '@koa/router';

import { characterRouter } from './characters';
import { healthCheckHandler } from './healthCheck';

export const router = new Router()
  .get('/health', healthCheckHandler)
  .use('/characters', characterRouter.routes());
