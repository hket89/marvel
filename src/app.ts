import { config } from './config';
import { errorHandler } from './lib/errors/errorHandler';
import { createApp } from './lib/server';
import { router } from './routes';
import { swagger } from './routes/swagger';

const app = createApp(errorHandler, router.allowedMethods(), router.routes());

const listener = app.listen(config.port, () => {
  const address = listener.address();

  if (typeof address === 'object' && address) {
    // eslint-disable-next-line no-console
    console.debug(`listening on port ${address.port}`);
  }
});

app.use(swagger);
export default listener;
