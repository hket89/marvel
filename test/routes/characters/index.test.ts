import { characterRouter } from '../../../src/routes/characters';
import { app } from '../../tools/app';
import { marvelSpy, mockData } from '../../tools/marvelApi';

describe('Router for character', () => {
  describe('GET /characters/:id', () => {
    afterEach(marvelSpy.clear);

    it('return status 400 on non id path', async () => {
      const { status } = await app(characterRouter).get('/fake');

      expect(status).toEqual(400);
    });

    it('return 200', async () => {
      marvelSpy.getCharacterById().mockResolvedValue(mockData);

      const result = await app(characterRouter).get('/123');

      expect(result).toMatchSnapshot();
    });

    it('return 404', async () => {
      marvelSpy
        .getCharacterById()
        .mockRejectedValue({ response: { status: 404 }, isAxiosError: true });

      const result = await app(characterRouter).get('/123');

      expect(result).toMatchSnapshot();
    });
  });

  describe('GET /characters', () => {
    afterEach(marvelSpy.clear);

    it('return 200', async () => {
      marvelSpy.getCharacters().mockResolvedValue(mockData);

      const result = await app(characterRouter).get('/');

      expect(result).toMatchSnapshot();
    });
  });
});
