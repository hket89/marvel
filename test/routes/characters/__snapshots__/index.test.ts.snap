// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router for character GET /characters return 200 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "7",
    "content-type": "application/json; charset=utf-8",
    "date": "Mon, 05 Apr 2021 18:24:31 GMT",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
    "url": "http://127.0.0.1:52138/",
  },
  "status": 200,
  "text": "[1,2,3]",
}
`;

exports[`Router for character GET /characters/:id return 200 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "42",
    "content-type": "application/json; charset=utf-8",
    "date": "Mon, 05 Apr 2021 18:24:31 GMT",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
    "url": "http://127.0.0.1:52134/123",
  },
  "status": 200,
  "text": "{\\"id\\":1,\\"name\\":\\"a\\",\\"description\\":\\"fake 1\\"}",
}
`;

exports[`Router for character GET /characters/:id return 404 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "2",
    "content-type": "text/plain; charset=utf-8",
    "date": "Mon, 05 Apr 2021 18:24:31 GMT",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
    "url": "http://127.0.0.1:52136/123",
  },
  "status": 404,
  "text": "{}",
}
`;
